src/commands/Invoke.ts

```diff
@@ -1,76 +1,122 @@
-import fetch, { BodyInit, Request, Response } from "node-fetch";
-import { ReadStream } from "fs";
-import queryString from "query-string";
-import { flatten } from "q-flat";
-import FormData from "form-data";
+import fetch, { Request, RequestInit, Response } from "node-fetch";
 
 import validateAndRefresh from "./RefreshIfNeeded";
 import retrieveImpersonatedCredentials from "./ImpersonateCredentials";
 
 import { Credentials, CredentialStorage } from "../oauth";
 import { ApiConfiguration } from "../ApiConfiguration";
+import queryString from 'query-string';
+import { flatten } from 'q-flat';
+import FormData from "form-data";
+
+export type AnyJson =  boolean | number | string | null | JsonArray | JsonMap;
+export interface JsonMap {  [key: string]: AnyJson; }
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface JsonArray extends Array<AnyJson> {}
+
+export type InvokeBody = Record<string, unknown> | [];
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type RequestResponse = Promise<any>;
+export type ResponseParser = (Response) => RequestResponse
+
+
+export class RequestBuilder {
+    private url: string
+    private init: RequestInit = {};
+    private responseParser: ResponseParser = async (response: Response) => response
+
+    constructor(url: string) {
+        this.url = url;
+    }
 
-export type InvokeBody = Record<string, unknown> | [] | ReadStream | Buffer;
+    public withMethod(method: string): RequestBuilder {
+        this.init.method = method
+        return this
+    }
+
+    public withConfig(config: ApiConfiguration): RequestBuilder {
+        this.init.agent = config.proxyAgent;
+        return this
+    }
+
+    public withHeader(header: string, value: string): RequestBuilder {
+        this.init.headers = { ...this.init.headers, [header]: value }
+        return this
+    }
+
+    public withFormData(payload?: Record<string, unknown> | []): RequestBuilder {
+        this.withHeader("Content-Type", "application/x-www-form-urlencoded")
+        this.init.body = queryString.stringify(flatten(payload), { encode: false })
+        return this
+    }
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
-const createRequest = (url: string, config: ApiConfiguration, credentials: Credentials, payload?: InvokeBody): Request => {
-    const headers = {
-        "Accept": "application/json",
-        "Authorization": `Bearer ${credentials.accessToken}`
-    };
-    let body: BodyInit | undefined;
-    if (url.endsWith('/trade-chat/image/upload') && payload) {
+    public withMultipartFormData(payload: Record<string, unknown> | []): RequestBuilder {
```

Comment:
Maybe better to add a type too here for `Record<string, unknown> | []`

Reply:
Replaced here and couple similar, thnx